# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [39]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [40]:
os.getcwd()

'/Users/shannonballard/Springboard Guided Capstone'

In [41]:
os.listdir()

['.DS_Store',
 'Notebook_stepthree.ipynb',
 'GuidedCapstoneStep3HL.ipynb',
 'Notebook_stepfourandfive.ipynb',
 'Step 2',
 '.ipynb_checkpoints',
 'Notebook_stepfour.ipynb']

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [109]:
df = pd.read_csv('/Users/shannonballard/Springboard Guided Capstone/Step 2/data/step3_output_b.csv')
df.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         0  
1               90.0             0.0         0  
2              152.0            30.0         0  
3              122.0             0.0         1  
4              104.0            80.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [110]:
#turn 'state' column into dummy variables
df_state = df['state']
dummies = pd.get_dummies(df_state)
dummies.head()

Alaska  Arizona  California  Colorado  Connecticut  Idaho  Illinois  \
0       1        0           0         0            0      0         0   
1       1        0           0         0            0      0         0   
2       1        0           0         0            0      0         0   
3       0        1           0         0            0      0         0   
4       0        1           0         0            0      0         0   

   Indiana  Iowa  Maine  ...  Rhode Island  South Dakota  Tennessee  Utah  \
0        0     0      0  ...             0             0          0     0   
1        0     0      0  ...             0             0          0     0   
2        0     0      0  ...             0             0          0     0   
3        0     0      0  ...             0             0          0     0   
4        0     0      0  ...             0             0          0     0   

   Vermont  Virginia  Washington  West Virginia  Wisconsin  Wyoming  
0        0         0           0              0          0        0  
1        0         0           0              0          0        0  
2        0         0           0              0          0        0  
3        0         0           0              0          0        0  
4        0         0           0              0          0        0  

[5 rows x 35 columns]

In [112]:
df2 = pd.concat([df, dummies], axis=1)
df2 = df2.drop('state', 1)
df2.head()

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0          0          2     2       0       0  ...             0   
1          0          0     0       0       4  ...             0   
2          0          0     0       1       0  ...             0   
3          1          0     2       2       1  ...             0   
4          0          1     2       3       1  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [113]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df2.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
 
y = df2['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [114]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [115]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [116]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [0]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [117]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

In [118]:
explained_variance_score(y_test, y_pred)

0.9403349390518856

In [119]:
mean_absolute_error(y_test, y_pred)

4.899680426186693

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [120]:
print(lm.intercept_)

64.10296277592712


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [121]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
# pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeff = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeff.head()

Coefficient
summit_elev    1.713602e+00
vertical_drop  1.082555e+00
trams          3.425078e+12
fastEight      3.367187e+11
fastSixes      3.986228e+12

In [122]:
sort = coeff.sort_values(['Coefficient'], ascending=[False])
top_ten = sort.head(10)
print(top_ten)

               Coefficient
total_chairs  3.546937e+13
fastQuads     1.344652e+13
surface       1.259838e+13
double        1.110216e+13
triple        9.903889e+12
quad          8.026736e+12
New York      4.509775e+12
Michigan      4.256003e+12
fastSixes     3.986228e+12
Colorado      3.749785e+12


**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [123]:
# Drop 'state' column
df3 = df.drop('state', 1)
df3.head()

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double  ...  SkiableTerrain_ac  \
0          0          2     2       0       0  ...             1610.0   
1          0          0     0       0       4  ...              640.0   
2          0          0     0       1       0  ...               30.0   
3          1          0     2       2       1  ...              777.0   
4          0          1     2       3       1  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         0  
1          53.0               90.0             0.0         0  
2          34.0              152.0            30.0         0  
3          89.0              122.0             0.0         1  
4          78.0              104.0            80.0         1  

[5 rows x 25 columns]

In [124]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X2 = df3.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
 
y2 = df3['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X2_scaled=scaler.transform(X2) 

In [125]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y2 = y2.ravel()

# Call the train_test_split() function with the first two parameters set to X2_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X2_train, X2_test, y2_train, y2_test = train_test_split(X2_scaled, y2, test_size=0.25, random_state=1)

In [126]:
#all second model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X2_train,y2_train)

In [127]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X2_test
y2_pred = model.predict(X2_test)

In [128]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

explained_variance_score(y2_test, y2_pred)

0.934761019889277

In [129]:
mean_absolute_error(y2_test, y2_pred)

5.126071445374538

In [130]:
print(lm.intercept_)

64.09509138559456


In [131]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
# pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeff = pd.DataFrame(abs(lm.coef_), X2.columns, columns=['Coefficient'])
coeff.head()

Coefficient
summit_elev       4.652246
vertical_drop     1.022142
trams             0.127082
fastEight         0.224541
fastSixes         0.280156

In [132]:
sort = coeff.sort_values(['Coefficient'], ascending=[False])
top_ten = sort.head(10)
print(top_ten)

                 Coefficient
AdultWeekday       19.841663
clusters            5.969742
summit_elev         4.652246
Runs                1.929555
surface             1.435298
triple              1.377548
averageSnowfall     1.081889
quad                1.074501
vertical_drop       1.022142
fastQuads           0.978870


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [133]:
# Drop 'summit_elev' column
df4 = df3.drop('summit_elev', 1)
df4.head()

Name  vertical_drop  trams  fastEight  fastSixes  fastQuads  \
0       Alyeska Resort           2500      1        0.0          0          2   
1  Eaglecrest Ski Area           1540      0        0.0          0          0   
2     Hilltop Ski Area            294      0        0.0          0          0   
3     Arizona Snowbowl           2300      0        0.0          1          0   
4  Sunrise Park Resort           1800      0        0.0          0          1   

   quad  triple  double  surface  ...  SkiableTerrain_ac  Snow Making_ac  \
0     2       0       0        2  ...             1610.0           113.0   
1     0       0       4        0  ...              640.0            60.0   
2     0       1       0        2  ...               30.0            30.0   
3     2       2       1        2  ...              777.0           104.0   
4     2       3       1        0  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         0  
1               90.0             0.0         0  
2              152.0            30.0         0  
3              122.0             0.0         1  
4              104.0            80.0         1  

[5 rows x 24 columns]

In [134]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X3,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X3 = df4.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y3, and assign it the AdultWeekend column of the df4 
 
y3 = df4['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X3 
scaler3 = preprocessing.StandardScaler().fit(X3)

# Declare a variable called X3_scaled, and assign it the result of calling the transform() method with parameter X3 
X3_scaled=scaler3.transform(X3) 

In [135]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y3 by calling the ravel() function on y3
y3 = y3.ravel()

# Call the train_test_split() function with the first two parameters set to X3_scaled and y3 
# Declare four variables, X3_train, X3_test, y3_train and y3_test separated by commas 
X3_train, X3_test, y3_train, y3_test = train_test_split(X3_scaled, y3, test_size=0.25, random_state=1)

In [136]:
#all second model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X3_train,y3_train)

In [137]:
# Make a variable called y3_pred and assign it the result of calling predict() on our model variable with parameter X3_test
y3_pred = model.predict(X3_test)

In [138]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

explained_variance_score(y3_test, y3_pred)

0.9354012974977128

In [139]:
mean_absolute_error(y3_test, y3_pred)

5.13034552797341

In [140]:
print(lm.intercept_)

64.07752725203471


In [141]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
# pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeff = pd.DataFrame(abs(lm.coef_), X3.columns, columns=['Coefficient'])
coeff.head()

Coefficient
vertical_drop     2.032411
trams             0.290860
fastEight         0.060005
fastSixes         0.383018
fastQuads         0.705187

In [142]:
sort = coeff.sort_values(['Coefficient'], ascending=[False])
top_ten = sort.head(10)
print(top_ten)

                  Coefficient
AdultWeekday        20.077469
clusters             2.638486
vertical_drop        2.032411
Runs                 1.398631
triple               1.387036
surface              1.249047
quad                 1.183929
daysOpenLastYear     1.156668
averageSnowfall      0.899505
fastQuads            0.705187


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9403 |  4.8997 |-|
| Model 2. | 0.9348| 5.1261 |'state'|
| Model 3. | 0.9354 | 5.1303 |'state','summit_elev','base_elev'|

Model Selection:

## Model 2
### want to choose the model based on the following criteria:
1. Lowest Mean Absolute Error (MAE)
2. Higher Explained Variance value

All 3 models were relatively close in their values for the above criteria, and Model 2 was chosen for further analysis.